packages/webpack-rtl-plugin/src/index.js

```diff
@@ -19,99 +19,118 @@ class WebpackRTLPlugin {
 	}
 
 	apply( compiler ) {
-		compiler.hooks.emit.tapAsync( pluginName, ( compilation, callback ) => {
-			forEachOfLimit(
-				compilation.chunks,
-				5,
-				( chunk, key, cb ) => {
-					const rtlFiles = [];
-					let cssnanoPromise = Promise.resolve();
-
-					chunk.files.forEach( ( asset ) => {
-						const match = this.options.test ? new RegExp( this.options.test ).test( asset ) : true;
-
-						if ( path.extname( asset ) !== '.css' ) {
-							return;
-						}
-
-						const baseSource = compilation.assets[ asset ].source();
-						let filename;
-						let rtlSource;
-
-						if ( match ) {
-							rtlSource = rtlcss.process( baseSource, this.options.options, this.options.plugins );
-
-							if ( this.options.filename instanceof Array && this.options.filename.length === 2 ) {
-								filename = asset.replace( this.options.filename[ 0 ], this.options.filename[ 1 ] );
-							} else if ( this.options.filename ) {
-								filename = this.options.filename;
-
-								if ( /\[contenthash]/.test( this.options.filename ) ) {
-									const hash = createHash( 'md5' )
-										.update( rtlSource )
-										.digest( 'hex' )
-										.substr( 0, 10 );
-									filename = filename.replace( '[contenthash]', hash );
+		compiler.hooks.thisCompilation.tap( pluginName, ( compilation ) => {
+			compilation.hooks.processAssets.tapAsync(
+				{ name: pluginName, stage: compilation.PROCESS_ASSETS_STAGE_DERIVED },
```

Comment:
I wasn't sure what's the practical difference between `PROCESS_ASSETS_STAGE_DERIVED` and `PROCESS_ASSETS_STAGE_ADDITIONAL`. While the choice here makes sense, do you have any idea of what is the difference between them in practice?

Reply:
I guess that different plugins now can tap into various stages and be orchestrated better, without conflicts. The stages are defined here: https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L3524

Adding a new `.rtl.css` asset derived from the `.css` one is a perfect use case for the `DERIVED` stage.

Ideally, we would postpone the `cssnano` processing until the `OPTIMIZE_SIZE` stage, in a separate hook. Or, even better, somehow move it out of the plugin completely.

Reply:
AFAIK, they define the order in which plugins for the same `processAssets` hook are run. They are defined in https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L3510.

Maybe I should use `PROCESS_ASSETS_STAGE_ADDITIONAL` so those new assets are included if any plugin does something on `PROCESS_ASSETS_STAGE_PRE_PROCESS`.

There are a couple of places where webpack itself uses `PROCESS_ASSETS_STAGE_ADDITIONAL`, but it never uses `PROCESS_ASSETS_STAGE_DERIVED`

Reply:
> Ideally, we would postpone the cssnano processing until the OPTIMIZE_SIZE stage, in a separate hook. Or, even better, somehow move it out of the plugin completely.

That's a good idea.

For now, all I'd like to change is to make the "original" (pre-fork) plugin compatible with webpack 5. We can work on extracting `cssnano` (and maybe apply to all CSS files, not only RTL's) later.

Reply:
> We can work on extracting `cssnano` (and maybe apply to all CSS files, not only RTL's) later.

Yes, it totally a material for a separate PR, not this one. The `cssnano` processing should be a part of the CSS loader chain, like it would be without RTL. But I never knew how to do it with this plugin that "forks" the CSS asset into two.